### DISCUSSION TITLE
Pot Ranges

### DISCUSSION THREAD

**Discussion by SimonG4ELI** [softerhardware/CWKeyer#13] [Q&A]
So,

Feb 12th, 07h48m looking through the code I don't see where the pot ranges are defined. Speed returns 5 to 54, others values in the 16k+ range. Really I'm looking for the logic for pot values other than Speed.

**Comment by recri** [comment]
In hasak/input.h: input_cook_value, sort of messy and overly tweaked.

-- rec --

**Comment by dl1ycf** [comment]
In the "CWKeyerShield" library and in hasak, this seems to be different. The CWKeyerShield library, which is used by TeensyWinkeyEmulator, produces values in the following ranges for a "full turn" of the four pots:

master volume:  0 - 127
side tone volume:  0 - 127
side tone frequency: 40-103, mapping to a frequency of 400 - 1030 Hz
CW speed:  5 - 52

In order to get rock-stable readings, the frequency pot is "discretized" into 64 different values, and
the speed pot into 32 different values. This is so to prevent these settings to be changed without the
pot actually turned.

The 64 different "bins" of freq pot readings are then interpreted as 400, 410, 420, ..., 1030 Hz
and the correponding MIDI values (40, 41, ..., 103) emitted. The 32 different "bins" of the speed
pot map to the actual speed by a table lookup, to have slower changes at the bottom and faster
changes at the upper end of the scale:

    uint8_t SpeedTab[32] = {  5,  6,  7,  8,  9, 10, 11, 12,
                             13, 14, 15, 16, 17, 18, 19, 20,
                             21, 22, 23, 24, 26, 28, 30, 32,
                             34, 36, 38, 40, 43, 46, 49, 52};

P.S. one could do the same for the frequency, having smaller delta-freq (in Hz) at the lower end and larger at the
upper end.

**Comment by recri** [comment]
Hasak is moving to linear interpolation of 10 bit analog read values into a
NRPN configurable segment of NRPN configurable 14 bit NRPN value, so you'll
be able to change the pot mapping range and slope as you find convenient.
Each analog pin is also going to get a samplePeriod.

The adc_read value is generated by the ResponsiveAnalogRead library.

Then you can switch the pots from [Volume, STVolume, Tone, Speed] to
[Speed, Weight, Ratio, Comp] or something else, and make them as coarse or
twitchy as you like.

So:
     adc_channel = { adc_pin, adc_nrpn, adc_period, adc_endpoint1,
adc_endpoint2 }
     if (pin_valid(adc_pin)) {
        every adc_period { nrpn_value[adc_nrpn] =
(adc_endpoint1*adc_read+adc_endpoint2*(1024-adc_read))/1024 }
    }

You can correct for forward/reverse wired pots by swapping the endpoints.

There will also be a way for some set of nrpn_values to map down to 7 bits
and fire midi notes for the incorrigible SDR apps.
